Add import hints for errors like 'Any' undefined

If new users forget to import `Any`, `List` or similar, it could be useful to add a hint such as this:

```
x.py:20: error: Name "Any" undefined
x.py:20: note: Did you forget to import it from "typing"? (Suggestion: "from typing import Any")
```

We could do this for some common stdlib functionality as well, such as `collections`, `subprocess`, `re`, etc.

Even better, mypy could also give import hints for some user-defined classes and functions (that have unique definitions).

--- a/mypy/newsemanal/semanal.py
+++ b/mypy/newsemanal/semanal.py
@@ -142,6 +142,21 @@ SUGGESTED_TEST_FIXTURES = {
     'builtins.classmethod': 'classmethod.pyi',
 }  # type: Final
 
+TYPES_FOR_UNIMPORTED_HINTS = {
+    'typing.Any',
+    'typing.Callable',
+    'typing.Dict',
+    'typing.Iterable',
+    'typing.Iterator',
+    'typing.List',
+    'typing.Optional',
+    'typing.Set',
+    'typing.Tuple',
+    'typing.TypeVar',
+    'typing.Union',
+    'typing.cast',
+}  # type: Final
+
 
 # Special cased built-in classes that are needed for basic functionality and need to be
 # available very early on.
@@ -3821,9 +3836,14 @@ class NewSemanticAnalyzer(NodeVisitor[None],
         return None
 
     def check_for_obsolete_short_name(self, name: str, ctx: Context) -> None:
-        matches = [obsolete_name
-                   for obsolete_name in obsolete_name_mapping
-                   if obsolete_name.rsplit('.', 1)[-1] == name]
+        lowercased_names_handled_by_unimported_hints = {
+            name.lower() for name in TYPES_FOR_UNIMPORTED_HINTS
+        }
+        matches = [
+            obsolete_name for obsolete_name in obsolete_name_mapping
+            if obsolete_name.rsplit('.', 1)[-1] == name
+            and obsolete_name not in lowercased_names_handled_by_unimported_hints
+        ]
         if len(matches) == 1:
             self.note("(Did you mean '{}'?)".format(obsolete_name_mapping[matches[0]]), ctx)
 
@@ -4296,12 +4316,32 @@ class NewSemanticAnalyzer(NodeVisitor[None],
             message += ' {}'.format(extra)
         self.fail(message, ctx)
         self.check_for_obsolete_short_name(name, ctx)
+
         if 'builtins.{}'.format(name) in SUGGESTED_TEST_FIXTURES:
             # The user probably has a missing definition in a test fixture. Let's verify.
             fullname = 'builtins.{}'.format(name)
             if self.lookup_fully_qualified_or_none(fullname) is None:
                 # Yes. Generate a helpful note.
                 self.add_fixture_note(fullname, ctx)
+
+        modules_with_unimported_hints = {
+            name.split('.', 1)[0]
+            for name in TYPES_FOR_UNIMPORTED_HINTS
+        }
+        lowercased = {
+            name.lower(): name
+            for name in TYPES_FOR_UNIMPORTED_HINTS
+        }
+        for module in modules_with_unimported_hints:
+            fullname = '{}.{}'.format(module, name).lower()
+            if fullname not in lowercased:
+                continue
+            # User probably forgot to import these types.
+            hint = (
+                'Did you forget to import it from "{module}"?'
+                ' (Suggestion: "from {module} import {name}")'
+            ).format(module=module, name=lowercased[fullname].rsplit('.', 1)[-1])
+            self.note(hint, ctx)
 
     def already_defined(self, name: str, ctx: Context,
                         original_ctx: Optional[Union[SymbolTableNode, SymbolNode]] = None, *,

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -166,8 +166,6 @@ SUGGESTED_TEST_FIXTURES = {
     'builtins.classmethod': 'classmethod.pyi',
 }  # type: Final
 
-# Map from the lowercased name of a missing definition to a tuple containing
-# the name of the likely module definition and a suggested import string.
 TYPES_FOR_UNIMPORTED_HINTS = {
     'typing.Any',
     'typing.Callable',

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -169,18 +169,18 @@ SUGGESTED_TEST_FIXTURES = {
 # Map from the lowercased name of a missing definition to a tuple containing
 # the name of the likely module definition and a suggested import string.
 TYPES_FOR_UNIMPORTED_HINTS = {
-    'any': ('typing', 'from typing import Any'),
-    'callable': ('typing', 'from typing import Callable'),
-    'dict': ('typing', 'from typing import Dict'),
-    'iterable': ('typing', 'from typing import Iterable'),
-    'iterator': ('typing', 'from typing import Iterator'),
-    'list': ('typing', 'from typing import List'),
-    'optional': ('typing', 'from typing import Optional'),
-    'set': ('typing', 'from typing import Set'),
-    'tuple': ('typing', 'from typing import Tuple'),
-    'typevar': ('typing', 'from typing import TypeVar'),
-    'union': ('typing', 'from typing import Union'),
-    'cast': ('typing', 'from typing import cast'),
+    'typing.Any',
+    'typing.Callable',
+    'typing.Dict',
+    'typing.Iterable',
+    'typing.Iterator',
+    'typing.List',
+    'typing.Optional',
+    'typing.Set',
+    'typing.Tuple',
+    'typing.TypeVar',
+    'typing.Union',
+    'typing.cast',
 }  # type: Final
 
 
@@ -3479,10 +3479,14 @@ class SemanticAnalyzerPass2(NodeVisitor[None],
         return None
 
     def check_for_obsolete_short_name(self, name: str, ctx: Context) -> None:
-        matches = [obsolete_name
-                   for obsolete_name in obsolete_name_mapping
-                   if obsolete_name.rsplit('.', 1)[-1] == name
-                   and name not in TYPES_FOR_UNIMPORTED_HINTS]
+        lowercased_names_handled_by_unimported_hints = {
+            name.lower() for name in TYPES_FOR_UNIMPORTED_HINTS
+        }
+        matches = [
+            obsolete_name for obsolete_name in obsolete_name_mapping
+            if obsolete_name.rsplit('.', 1)[-1] == name
+            and obsolete_name not in lowercased_names_handled_by_unimported_hints
+        ]
         if len(matches) == 1:
             self.note("(Did you mean '{}'?)".format(obsolete_name_mapping[matches[0]]), ctx)
 
@@ -3760,13 +3764,23 @@ class SemanticAnalyzerPass2(NodeVisitor[None],
                 # Yes. Generate a helpful note.
                 self.add_fixture_note(fullname, ctx)
 
-        if name.lower() in TYPES_FOR_UNIMPORTED_HINTS:
+        modules_with_unimported_hints = {
+            name.split('.', 1)[0]
+            for name in TYPES_FOR_UNIMPORTED_HINTS
+        }
+        lowercased = {
+            name.lower(): name
+            for name in TYPES_FOR_UNIMPORTED_HINTS
+        }
+        for module in modules_with_unimported_hints:
+            fullname = '{}.{}'.format(module, name).lower()
+            if fullname not in lowercased:
+                continue
             # User probably forgot to import these types.
-            unimported_module, suggestion = TYPES_FOR_UNIMPORTED_HINTS[name.lower()]
             hint = (
-                f'Did you forget to import it from "{unimported_module}"?'
-                f' (Suggestion: "{suggestion}")'
-            )
+                'Did you forget to import it from "{module}"?'
+                ' (Suggestion: "from {module} import {name}")'
+            ).format(module=module, name=lowercased[fullname].rsplit('.', 1)[-1])
             self.note(hint, ctx)
 
     def already_defined(self, name: str, ctx: Context,

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -3481,7 +3481,8 @@ class SemanticAnalyzerPass2(NodeVisitor[None],
     def check_for_obsolete_short_name(self, name: str, ctx: Context) -> None:
         matches = [obsolete_name
                    for obsolete_name in obsolete_name_mapping
-                   if obsolete_name.rsplit('.', 1)[-1] == name]
+                   if obsolete_name.rsplit('.', 1)[-1] == name
+                   and name not in TYPES_FOR_UNIMPORTED_HINTS]
         if len(matches) == 1:
             self.note("(Did you mean '{}'?)".format(obsolete_name_mapping[matches[0]]), ctx)
 

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -166,6 +166,23 @@ SUGGESTED_TEST_FIXTURES = {
     'builtins.classmethod': 'classmethod.pyi',
 }  # type: Final
 
+# Map from the lowercased name of a missing definition to a tuple containing
+# the name of the likely module definition and a suggested import string.
+TYPES_FOR_UNIMPORTED_HINTS = {
+    'any': ('typing', 'from typing import Any'),
+    'callable': ('typing', 'from typing import Callable'),
+    'dict': ('typing', 'from typing import Dict'),
+    'iterable': ('typing', 'from typing import Iterable'),
+    'iterator': ('typing', 'from typing import Iterator'),
+    'list': ('typing', 'from typing import List'),
+    'optional': ('typing', 'from typing import Optional'),
+    'set': ('typing', 'from typing import Set'),
+    'tuple': ('typing', 'from typing import Tuple'),
+    'typevar': ('typing', 'from typing import TypeVar'),
+    'union': ('typing', 'from typing import Union'),
+    'cast': ('typing', 'from typing import cast'),
+}  # type: Final
+
 
 class SemanticAnalyzerPass2(NodeVisitor[None],
                             SemanticAnalyzerInterface,
@@ -3734,12 +3751,22 @@ class SemanticAnalyzerPass2(NodeVisitor[None],
         if extra:
             message += ' {}'.format(extra)
         self.fail(message, ctx)
+
         if 'builtins.{}'.format(name) in SUGGESTED_TEST_FIXTURES:
             # The user probably has a missing definition in a test fixture. Let's verify.
             fullname = 'builtins.{}'.format(name)
             if self.lookup_fully_qualified_or_none(fullname) is None:
                 # Yes. Generate a helpful note.
                 self.add_fixture_note(fullname, ctx)
+
+        if name.lower() in TYPES_FOR_UNIMPORTED_HINTS:
+            # User probably forgot to import these types.
+            unimported_module, suggestion = TYPES_FOR_UNIMPORTED_HINTS[name.lower()]
+            hint = (
+                f'Did you forget to import it from "{unimported_module}"?'
+                f' (Suggestion: "{suggestion}")'
+            )
+            self.note(hint, ctx)
 
     def already_defined(self, name: str, ctx: Context,
                         original_ctx: Optional[Union[SymbolTableNode, SymbolNode]] = None, *,

--- a/test-data/unit/check-expressions.test
+++ b/test-data/unit/check-expressions.test
@@ -2206,3 +2206,21 @@ C in exp  # E: Non-overlapping container check (element type: "Type[C]", contain
 o in exp
 [builtins fixtures/list.pyi]
 [typing fixtures/typing-full.pyi]
+
+
+[case testUnimportedHintAny]
+def f(x: Any) -> None:  # E: Name 'Any' is not defined \
+                        # N: Did you forget to import it from "typing"? (Suggestion: "from typing import Any")
+    pass
+
+
+[case testUnimportedHintAnyLower]
+def f(x: any) -> None:  # E: Name 'any' is not defined \
+                        # N: Did you forget to import it from "typing"? (Suggestion: "from typing import Any")
+    pass
+
+
+[case testUnimportedHintOptional]
+def f(x: Optional[str]) -> None:  # E: Name 'Optional' is not defined \
+                                  # N: Did you forget to import it from "typing"? (Suggestion: "from typing import Optional")
+    pass

--- a/test-data/unit/check-inference.test
+++ b/test-data/unit/check-inference.test
@@ -1553,7 +1553,8 @@ def add():
 [builtins fixtures/dict.pyi]
 
 [case testSpecialCaseEmptyListInitialization]
-def f(blocks: Any): # E: Name 'Any' is not defined
+def f(blocks: Any): # E: Name 'Any' is not defined \
+                    # N: Did you forget to import it from "typing"? (Suggestion: "from typing import Any")
     to_process = [] # E: Need type annotation for 'to_process'
     to_process = list(blocks)
 [builtins fixtures/list.pyi]

--- a/test-data/unit/fine-grained.test
+++ b/test-data/unit/fine-grained.test
@@ -7841,11 +7841,13 @@ def bar(x: T) -> T:
 x = 1
 [out]
 a.py:1: error: Name 'TypeVar' is not defined
+a.py:1: note: Did you forget to import it from "typing"? (Suggestion: "from typing import TypeVar")
 a.py:7: error: Invalid type "a.T"
 a.py:10: error: Name 'bar' already defined on line 6
 a.py:11: error: Invalid type "a.T"
 ==
 a.py:1: error: Name 'TypeVar' is not defined
+a.py:1: note: Did you forget to import it from "typing"? (Suggestion: "from typing import TypeVar")
 a.py:7: error: Invalid type "a.T"
 a.py:10: error: Name 'bar' already defined on line 6
 a.py:11: error: Invalid type "a.T"

--- a/test-data/unit/pythoneval.test
+++ b/test-data/unit/pythoneval.test
@@ -242,6 +242,7 @@ import typing
 cast(int, 2)
 [out]
 _program.py:2: error: Name 'cast' is not defined
+_program.py:2: note: Did you forget to import it from "typing"? (Suggestion: "from typing import cast")
 
 [case testBinaryIOType]
 from typing import BinaryIO

--- a/test-data/unit/semanal-errors.test
+++ b/test-data/unit/semanal-errors.test
@@ -1220,6 +1220,7 @@ def f() -> List[int]: pass
 [builtins fixtures/list.pyi]
 [out]
 main:2: error: Name 'List' is not defined
+main:2: note: Did you forget to import it from "typing"? (Suggestion: "from typing import List")
 
 [case testImportObsoleteTypingFunction]
 from typing import Function # E: Module 'typing' has no attribute 'Function' (it's now called 'typing.Callable')
@@ -1260,7 +1261,7 @@ main:1: error: Module 'typing' has no attribute 'typevar' (it's now called 'typi
 t = typevar('t')
 [out]
 main:1: error: Name 'typevar' is not defined
-main:1: note: (Did you mean 'typing.TypeVar'?)
+main:1: note: Did you forget to import it from "typing"? (Suggestion: "from typing import TypeVar")
 
 [case testUseObsoleteNameForTypeVar3]
 import typing

